lib/ansible/galaxy/collection.py

```diff
@@ -0,0 +1,265 @@
+                manifest_entry['name'] = rel_path
+                manifest_entry['ftype'] = 'dir'
+
+                manifest.append(manifest_entry)
+
+                _walk(abs_path, top_level_dir, parent_dirs)
+            else:
+                if item == 'galaxy.yml':
+                    continue
+                elif any(fnmatch.fnmatch(item, pattern) for pattern in ignore_files):
+                    display.vvv("Skipping %s for collection build" % abs_path)
+                    continue
+
+                manifest_entry = entry_template.copy()
+                manifest_entry['name'] = rel_path
+                manifest_entry['ftype'] = 'file'
+                manifest_entry['chksum_type'] = 'sha256'
+                manifest_entry['chksum_sha256'] = secure_hash(abs_path, hash_func=sha256)
+
+                manifest.append(manifest_entry)
+
+    dir_stats = os.stat(collection_path)
+    parents = frozenset(((dir_stats.st_dev, dir_stats.st_ino),))
+    _walk(collection_path, collection_path, parents)
+
+    return manifest
+
+
+def _build_manifest(namespace, name, version, authors, tags, description, license_ids, dependencies, **kwargs):
+
+    def _parse_author_info(author):
+        email = None
+        url = None
+
+        match = AUTHOR_REGEX.match(author)
+        if match:
+            author = match.group('name')
+            email = match.group('email')
+            url = match.group('url')
+
+        return {
+            'name': author,
+            'email': email,
+            'url': url
+        }
+    authors = [_parse_author_info(info) for info in authors]
+
+    deps = {}
+    for name, version in dependencies.items():
+        deps[name] = {'version': version}
+
+    manifest = {
+        'format': MANIFEST_FORMAT,
+        'name': "%s.%s" % (namespace, name),
+        'version': version,
+        'authors': authors,
+        'keywords': tags,
+        'description': description,
+        'license': license_ids,
+        'license_file': None,  # TODO: thaumos to verify the need for license_file and readme
+        'readme': None,
+        'dependencies': deps,
+        'file_manifest_file': {
+            'name': 'FILES.json',
+            'chksum_type': 'sha256',
+            'chksum_sha256': None,  # Filled out in _build_collection_tar
+        },
+    }
+
+    return manifest
+
+
+def _build_collection_tar(collection_path, tar_path, collection_manifest, file_manifest):
+    files_manifest_json = to_bytes(json.dumps(file_manifest), errors='surrogate_or_strict')
+    collection_manifest['file_manifest_file']['chksum_sha256'] = secure_hash_s(files_manifest_json, hash_func=sha256)
+    collection_manifest_json = to_bytes(json.dumps(collection_manifest), errors='surrogate_or_strict')
+
+    tempdir = tempfile.mkdtemp(dir=C.DEFAULT_LOCAL_TMP)
+    try:
+        tar_filepath = os.path.join(tempdir, os.path.basename(tar_path))
+        tar_file = tarfile.open(tar_filepath, mode='w:gz')
```

Comment:
Plz use a context manager to have a safe resource closing
```suggestion
        with tarfile.open(tar_filepath, mode='w:gz') as tar_file:
```

Reply:
As per the talk offline the context manager for tarfile was only added in Python 2.7. While we don't support it we still run unit tests with 2.6. I've created our own context manager to handle this.

Reply:
Where's it? I cannot see it.

Reply:
See https://github.com/ansible/ansible/pull/57106#discussion_r294229128